### PR TITLE
test: added a new test to ensure null prefix works

### PR DESCRIPTION
**Pull request by kierramarie**
### Description

pr_test.go was updated to include a new test that ensures the prefix can be set to null. A new example was added to be used in this test. The example has the prefix hard-coded as null.

### Release required?

- [x] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- When merging, use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: examples/no-prefix/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default Example
```
Comment (ocofaigh) [inline examples/no-prefix/README.md:1]: update this title so its different from the actual default example

File: examples/no-prefix/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+
+An end to end example using the module's default variable values.
```
Comment (ocofaigh) [inline examples/no-prefix/README.md:3]: update description with more info on what this example shows

File: examples/no-prefix/variables.tf [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    error_message = "Keys for `use_public_gateways` must be in the order `zone-1`, `zone-2`, `zone-3`."
+    condition     = var.address_prefixes == null ? true : (keys(var.address_prefixes)[0] == "zone-1" && keys(var.address_prefixes)[1] == "zone-2" && keys(var.address_prefixes)[2] == "zone-3")
+  }
+}
+
+variable "network_cidrs" {
+  description = "List of Network CIDRs for the VPC. This is used to manage network ACL rules for cluster provisioning."
+  type        = list(string)
+  default     = ["10.0.0.0/8", "164.0.0.0/8"]
+}
```
Comment (ocofaigh) [inline examples/no-prefix/variables.tf:93]: can we remove all these as variables and add them directly into the main.tf. It makes it easier for consumers to see what kind of values need to be passed for different vars, but it also prevents people changing variable values which may break the example.
(This is something we need to roll out to all repo examples eventually)

File: examples/no-prefix/README.md [hunk 4]
```diff
@@ -1,3 +1,3 @@
-# Default Example
+# No Prefix Example
 
-An end to end example using the module's default variable values.
+An example similar to the basic example except in this example the prefix is hard-coded to be null. This example is used in a PR test.
```
Comment (ocofaigh) [inline examples/no-prefix/README.md:3]: there is no basic example - its called default. Remove the line `This example is used in a PR test.`

File: examples/no-prefix/variables.tf (hunk 3 above)
Comment (ocofaigh) [reply examples/no-prefix/variables.tf:93]: @kierramarie still outstanding

Comment (kierramarie) [reply examples/no-prefix/variables.tf:93]: which variables do you want to remain in the variables.tf?

File: tests/pr_test.go [hunk 5]
```diff
@@ -54,6 +57,29 @@ func TestRunBasicExample(t *testing.T) {
 	assert.NotNil(t, output, "Expected some output")
 }
 
+func TestRunNull(t *testing.T) {
```
Comment (toddgiguere) [inline tests/pr_test.go:60]: I think we should change this test name to `TestRunNoPrefixExample`

File: examples/no-prefix/main.tf [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+#############################################################################
+# Provision VPC
+#############################################################################
+
+module "slz_vpc" {
+  source                                 = "../../"
+  resource_group_id                      = var.resource_group != null ? data.ibm_resource_group.existing_resource_group[0].id : ibm_resource_group.resource_group[0].id
+  region                                 = var.region
+  name                                   = var.name
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:46]: don't we now want to update this to `name = var.prefix` so our tests will pass a unique prefix to it?

**Review by toddgiguere — CHANGES_REQUESTED** [review]

File: examples/no-prefix/main.tf (hunk 6 above)
Comment (ocofaigh) [reply examples/no-prefix/main.tf:46]: ah actually never mind, we are doing this directly in the test

File: examples/no-prefix/variables.tf [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+variable "region" {
+  description = "The region to which to deploy the VPC"
+  type        = string
+  default     = "us-south"
+}
+
+variable "prefix" {
+  description = "The prefix that you would like to append to your resources"
```
Comment (ocofaigh) [inline examples/no-prefix/variables.tf:14]: well this is not actually true - its used for all resources except the ones created in the VPC module itself.

File: examples/no-prefix/main.tf (hunk 6 above)
Comment (kierramarie) [reply examples/no-prefix/main.tf:46]: The prefix should be null. We want to pass a unique name to it.

Comment (toddgiguere) [reply examples/no-prefix/main.tf:46]: @ocofaigh in the pr_test code there is a different unique value being generated for the name, and it is passed into the name variable for the test and checked afterwards. The prefix and name will use different and random/unique values.

Done this way because the unit test doesn't have direct access to the unique value used as part of the prefix, and the entire prefix is already a bit long.

File: examples/no-prefix/variables.tf (hunk 3 above)
Comment (toddgiguere) [reply examples/no-prefix/variables.tf:93]: and `name`, we still need the name input for this particular example (see other comment about how we are using that input in pr_test)

File: examples/no-prefix/main.tf [hunk 8]
```diff
@@ -14,21 +14,38 @@ data "ibm_resource_group" "existing_resource_group" {
...
   service           = "cloud-object-storage"
-  plan              = var.cos_plan
-  location          = var.cos_location
+  plan              = local.cos_plan
+  location          = local.cos_location
 }
 
 resource "ibm_cos_bucket" "cos_bucket" {
-  count                = var.enable_vpc_flow_logs ? 1 : 0
+  count                = local.enable_vpc_flow_logs ? 1 : 0
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:48]: no need for a count now that we are hard coding `enable_vpc_flow_logs` to `true`

File: examples/no-prefix/main.tf [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  count = var.resource_group != null ? 1 : 0
+  name  = var.resource_group
+}
+
+#############################################################################
+# Provision cloud object storage and bucket
+#############################################################################
+
+resource "ibm_resource_instance" "cos_instance" {
+  count             = 1
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:22]: you dont need this count

File: examples/no-prefix/main.tf [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  name              = "${var.prefix}-vpc-logs-cos"
+  resource_group_id = var.resource_group != null ? data.ibm_resource_group.existing_resource_group[0].id : ibm_resource_group.resource_group[0].id
+  service           = "cloud-object-storage"
+  plan              = "standard"
+  location          = "global"
+}
+
+resource "ibm_cos_bucket" "cos_bucket" {
+  count                = 1
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:31]: you dont need this count

File: examples/no-prefix/main.tf [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  source                                 = "../../"
+  resource_group_id                      = var.resource_group != null ? data.ibm_resource_group.existing_resource_group[0].id : ibm_resource_group.resource_group[0].id
+  region                                 = var.region
+  name                                   = var.name
+  prefix                                 = null
+  tags                                   = var.resource_tags
+  access_tags                            = []
+  enable_vpc_flow_logs                   = true
+  create_authorization_policy_vpc_to_cos = true
+  existing_cos_instance_guid             = ibm_resource_instance.cos_instance[0].guid
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:52]: after removing the count, this can be updated to `ibm_resource_instance.cos_instance.guid`

File: examples/no-prefix/main.tf [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  existing_storage_bucket_name           = ibm_cos_bucket.cos_bucket[0].bucket_name
```
Comment (ocofaigh) [inline examples/no-prefix/main.tf:53]: after removing the count, this can be updated to `ibm_cos_bucket.cos_bucket.bucket_name`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
see comments on how to remove count

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Looks like we need to be more specific with our unique prefxi generation. It can't start with a number:

```
TestRunNoPrefixExample 2023-09-07T09:11:36Z logger.go:66: │ Error: "name" ("9cyinv-vpc") should match regexp ^([a-z]|[a-z][-a-z0-9]*[a-z0-9])$ 
TestRunNoPrefixExample 2023-09-07T09:11:36Z logger.go:66: │ 
TestRunNoPrefixExample 2023-09-07T09:11:36Z logger.go:66: │   with module.slz_vpc.ibm_is_vpc.vpc,
TestRunNoPrefixExample 2023-09-07T09:11:36Z logger.go:66: │   on ../../main.tf line 6, in resource "ibm_is_vpc" "vpc":
TestRunNoPrefixExample 2023-09-07T09:11:36Z logger.go:66: │    6:   name                        = var.prefix != null ? "${var.prefix}-${var.name}-vpc" : "${var.name}-vpc"
```

**Comment by toddgiguere** [comment]
@kierramarie We can add a small prefix string to it, to ensure it starts with a letter.

**Review by toddgiguere — APPROVED** [review]
